File: java/src/main/api/Boundr.java
package main.api;

import main.config.BoundrConfig;
import main.registry.BucketRegistry;
import main.time.SystemTimeProvider;
import main.time.TimeProvider;

/* Public API for rate limiting */

public class Boundr {
    private final BucketRegistry registry;

    public Boundr(BoundrConfig config) {
        this(config, new SystemTimeProvider());
    };

    public Boundr(BoundrConfig config, TimeProvider timeProvider) {
        this.registry = new BucketRegistry(config.getCapacity(), config.getRefillInterval(), timeProvider);
    };

    /* @return true if request is allowed, false otherwise */

    public boolean allow(String key) {
        if (key == null || key.isEmpty()) {
            throw new IllegalArgumentException("Key must not be null");
        }
        return registry.getBucket(key).consumeToken();
    };

    /* Cleanup of unused buckets */

    public void cleanupBuckets() {
        registry.cleanupUnusedBuckets();
    };
}

File: typescript/src/api/Boundr.ts
import { BoundrConfig } from "../config/BoundrConfig";
import { BucketRegistry } from "../registry/BucketRegistry";
import { SystemTimeProvider } from "../time/SystemTimeProvider";
import { TimeProvider } from "../time/TimeProvider";

/* Public API for rate limiting */

export class Boundr {
  private readonly registry: BucketRegistry;

  constructor(config: BoundrConfig, timeProvider?: TimeProvider) {
    const time = timeProvider ?? new SystemTimeProvider();

    this.registry = new BucketRegistry(
      config.getCapacity(),
      config.getRefillInterval(),
      time
    );
  }

  /* @return true if request is allowed, false otherwise */

  public allow(key: string): boolean {
    if (!key || key.trim().length === 0) {
      throw new Error("Key must not be empty");
    }
    return this.registry.getBucket(key).consumeToken();
  }

  /* Cleanup of unused buckets */

  public cleanupBuckets(): void {
    this.registry.cleanupUnusedBuckets();
  }
}
